src/main/java/use_cases/attack_use_case/AttackInteractor.java

```diff
@@ -0,0 +1,121 @@
+            attackPresenter.showDefendInputScreen(attackModel);
+        }
+
+    }
+
+    /**
+     * Process the attack using the attacker and target in the AttackRequestModel.
+     * The target in this AttackRequestModel must be the final target to receive
+     * the attack.
+     *
+     * @param inputData the AttackRequestModel containing the attacking creature's
+     *                  id, and the target creature's id.
+     * @return the FinishAttackResponseModel reporting the updates to the stats
+     * of creatures and stats of players in the game.
+     */
+    @Override
+    public FinishAttackResponseModel proceedAttack(AttackRequestModel inputData) {
+        Player player1 = gameState.getCurrentPlayer();
+        Player player2 = gameState.getOpposingPlayer();
+        CreatureCard creatureAttacker = player1.getCreatureById(inputData.getAttackerId());
+        CreatureCard creatureDefender = player2.getCreatureById(inputData.getTargetId());
+        player1.spendEssence(creatureAttacker.getAttackCost());
+        return getFinishAttackResponseModel(player1, player2, creatureAttacker, creatureDefender);
+
+    }
+
+    @Override
+    public FinishAttackResponseModel defend(AttackRequestModel inputData) {
+        Player player1 = gameState.getCurrentPlayer();
+        Player player2 = gameState.getOpposingPlayer();
+        CreatureCard creatureAttacker = player1.getCreatureById(inputData.getAttackerId());
+        CreatureCard creatureDefender = player2.getCreatureById(inputData.getTargetId());
+
+        player1.spendEssence(creatureAttacker.getAttackCost());
+        player2.spendEssence(creatureDefender.getDefendCost());
+        return getFinishAttackResponseModel(player1, player2, creatureAttacker, creatureDefender);
+
+    }
+
+    private FinishAttackResponseModel getFinishAttackResponseModel(Player player1, Player player2, CreatureCard
+            creatureAttacker, CreatureCard creatureDefender) {
+
+        creatureDefender.takeDamage(creatureAttacker.getAttackDamage());
+
+        List<List<Integer>> player1HitIdHand= getPlayerHitCreatureHandID(player1);
+        List<List<Integer>> player2HitIdHand= getPlayerHitCreatureHandID(player2);
+
+        List<Integer> player1CreatureHitpoints = player1HitIdHand.get(0);
+        List<Integer> player2CreatureHitpoints = player2HitIdHand.get(0);
+        List<Integer>  player1CreatureIds = player1HitIdHand.get(1);
+        List<Integer> player2CreatureIds = player1HitIdHand.get(1);
+        List <Integer> player1HandIds = player1HitIdHand.get(2);
+        List <Integer> player2HandIds = player1HitIdHand.get(2);
+
+        FinishAttackResponseModel updatedModel = new FinishAttackResponseModel(player1CreatureHitpoints,
+                player2CreatureHitpoints, player1CreatureIds, player2CreatureIds, player1HandIds, player2HandIds);
+
+        return attackPresenter.exitDefendInputScreen(updatedModel);
+    }
```

Comment:
This helper method is reasonable, so let's keep it. But as I just commented on the code below this one, it's not necessary to use another helper function `getPlayerHitCreatureHandID`, it can get overly complicated.

Instead, just do this:
(1) Set up a few for loops. We can just iterate through the creatures as we append the required stats to the lists of stats. And we can just iterate through the player's hand to collect the list of hand card ids.
(2) Some changes to the naming of the variables, keeping it consistent with the names in the data model.
(3) Some formatting, you can see that lines 87 to 92 are not consistently spaced. It's always good to do auto formatting with the IDE, but we can have a branch later on to specifically fix styles and conventions. (On Mac, IntelliJ's auto format shortcut is **Option + Command + L**).
```suggestion
    private FinishAttackResponseModel getFinishAttackResponseModel(Player player1, Player player2,
                                                                   CreatureCard creatureAttacker,
                                                                   CreatureCard creatureDefender) {
        creatureDefender.takeDamage(creatureAttacker.getAttackDamage());
        
        List<Integer> hitPoints1 = new ArrayList<>();
        List<Integer> hitPoints2 = new ArrayList<>();
        List<Integer> creatureIds1 = new ArrayList<>();
        List<Integer> creatureIds2 = new ArrayList<>();
        List<Integer> handIds1 = new ArrayList<>();
        List<Integer> handIds2 = new ArrayList<>();

        for (CreatureCard creature : player1.getCreatures()) {
            hitPoints1.add(creature.getTotalHitPoints());
            creatureIds1.add(creature.getId());
        }

        for (CreatureCard creature : player2.getCreatures()) {
            hitPoints2.add(creature.getTotalHitPoints());
            creatureIds2.add(creature.getId());
        }

        for (Card card : player1.getHand()) {
            handIds1.add(card.getId());
        }

        for (Card card : player2.getHand()) {
            handIds2.add(card.getId());
        }

        FinishAttackResponseModel updatedModel = new FinishAttackResponseModel(hitPoints1,
                hitPoints2, creatureIds1, creatureIds2, handIds1, handIds2);

        return attackPresenter.exitDefendInputScreen(updatedModel);
    }
```